daemon/containerd/image_exporter.go

```diff
@@ -58,49 +59,101 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
 		archive.WithPlatform(platform),
 	}
 
-	ctx, release, err := i.client.WithLease(ctx)
-	if err != nil {
-		return errdefs.System(err)
-	}
-	defer release(ctx)
-
 	for _, name := range names {
 		target, err := i.resolveDescriptor(ctx, name)
 		if err != nil {
 			return err
 		}
 
-		// We may not have locally all the platforms that are specified in the index.
-		// Export only those manifests that we have.
-		// TODO(vvoland): Reconsider this when `--platform` is added.
-		if containerdimages.IsIndexType(target.MediaType) {
-			desc, err := i.getBestDescriptorForExport(ctx, target)
-			if err != nil {
-				return err
+		var cleanup func()
+		if cleanup, err = i.leaseContent(ctx, target, func() error {
```

Comment:
Very clever. Too clever, IMO. All the outstanding leases are unconditionally cleaned up when `ExportImage` returns so it makes no difference whether the lease cleanup is deferred before or after the code inside the closure.
```go
lease, err := i.leaseContent(ctx, target)
if err != nil {
    return err
}
defer func(l containerd.Lease) {
	if derr := ls.Delete(ctx, l); derr != nil {
		log.G(ctx).WithError(derr).Warn("cleaning up lease")
	}
}(lease)
if containerdimages.IsIndexType(target.MediaType) {
	/* ... */
}
/* ... */
```

Furthermore, why create a separate lease for each image? Couldn't a single lease be used that covers the whole export operation?
```go
func (i *ImageService) leaseContent(ctx context.Context, lease containerd.Lease, desc ocispec.Descriptor) error
```

Reply:
Done